configuration/klippy/beacon_adaptive_heat_soak.py

```
@@ -0,0 +1,722 @@
+				while True:
+					if self.reactor.monotonic() - start_time > maximum_wait:
+						gcmd.respond_info(f"Maximum wait time of {self._format_seconds(maximum_wait)} exceeded, wait completed.")
+						return
+
+					try:
+						z_rate_result = z_rate_session.get_next_z_rate()
+					except Exception as e:
+						if self.printer.is_shutdown():
+							raise
+						else:
+							raise self.printer.command_error(f"Error calculating Z-rate, wait ended prematurely: {e}")
```

Ruff (0.12.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)